packages/botfuel-dialog/src/dialog-manager.js

```diff
@@ -238,46 +247,56 @@ class DialogManager extends Resolver {
         data: {},
       });
     } else {
-      const action = await this.resolve(dialog.name).execute(adapter, userMessage, dialog.data);
+      const { action, botMessages: newBotMessages } = await this.resolve(dialog.name).execute(
+        userMessage,
+        dialog.data,
+      );
+      botMessages = botMessages.concat(newBotMessages);
+
       logger.debug('execute: action', action);
       if (action.name !== Dialog.ACTION_WAIT) {
         dialogs = await this.applyAction(dialogs, action);
       } else {
-        return dialogs;
+        return {
+          dialogs,
+          botMessages,
+        };
       }
     }
-    return this.execute(adapter, userMessage, dialogs);
+    return this.execute(userMessage, dialogs, botMessages);
   }
 
   /**
    * Executes when receiving the classification results and message entities.
-   * @param {Adapter} adapter - the adapter
    * @param {Object} userMessage - the user message
    * @param {Object[]} classificationResults - the classification results from trainer
    * @param {Object[]} messageEntities - the message entities
    * @returns {void}
    */
-  async executeClassificationResults(adapter, userMessage, classificationResults, messageEntities) {
+  async executeClassificationResults(userMessage, classificationResults, messageEntities) {
     logger.debug('classificationResult', userMessage, classificationResults, messageEntities);
     const userId = userMessage.user;
     const dialogs = await this.getDialogs(userId);
     this.updateWithClassificationResults(userId, dialogs, classificationResults, messageEntities);
-    await this.setDialogs(userId, await this.execute(adapter, userMessage, dialogs));
+    const { dialogs: newDialogs, botMessages } = await this.execute(userMessage, dialogs, []);
+    await this.setDialogs(userId, newDialogs);
+    return botMessages;
   }
 
   /**
    * Populates and executes the stack.
-   * @param {Adapter} adapter - the adapter
    * @param {Object} userMessage - the user message
    * @param {Object[]} newDialog - the new dialogs
    * @returns {void}
    */
-  async executeDialog(adapter, userMessage, newDialog) {
+  async executeDialog(userMessage, newDialog) {
     logger.debug('executeDialog', userMessage, newDialog);
     const userId = userMessage.user;
     const dialogs = await this.getDialogs(userId);
     this.updateWithDialog(dialogs, newDialog);
-    await this.setDialogs(userId, await this.execute(adapter, userMessage, dialogs));
+    const { dialogs: newDialogs, botMessages } = await this.execute(userMessage, dialogs, []);
```

Comment:
maybe introduce a method to factorize these lines

Reply:
I find no trivial way (because I think that `getDialogs` and `setDialogs` should be called on the same function for symmetry). We should discuss about it. I created a task on asana: https://app.asana.com/0/370953015396593/638337161683023